### ISSUE TITLE
Findings trying to make macOS build again

### ISSUE THREAD

**Issue by JesperTreetop** [opensludge/opensludge#10]
I checked in on SLUDGE and saw that it hadn't gotten any attention on Mac since a few years back, and I sat down trying to see if I could bring it up to speed. (The 2019 release of macOS will drop support for 32-bit binaries, so there's reason to produce a build again soon.)

These are a few of the issues I encountered. Most of them are just the platform having changed since the last time - I simply mean to document them to help anyone giving this a try after me. Ultimately it looked like these changes would be too disruptive to the current state to be done by an outsider, so I thought the best I could do was to write it up instead.

* It is built against the Universal Mac OS X 10.4 SDK, which is long since unsupported. I changed it to the most current SDK.
* It builds using GCC, which is similarly not provided. clang is used on macOS nowadays (referred in Xcode as the "LLVM compiler"); I changed it to use that.
* GLee is not included in the repository, probably intended to be provided by the developer. The steps to build for Linux include how to retrieve it, but there are no steps for building on Mac and the project file assumes a `.c` file is included as well as the header, so it's intended to be provided manually? Trying to retrieve it from its site reveals that [the site is busted](http://elf-stone.com/glee.php). Grabbing it from [a random other repository](https://github.com/Studiofreya/code-samples/blob/master/opengl/collisiondetect/GLee.c) worked.
* Carbon is referred to by being a linked framework and in GLee.c. Replacing it with `#include <CoreFoundation/CoreFoundation.h>` worked. (GLee only uses it for the CoreFoundation APIs.)
* libpng.framework is referred to and used. It is no longer a framework provided by the SDK or system, and [recent libpng downloads](http://www.libpng.org/pub/png/libpng.html) no longer provide an Xcode project, and therefore not a libpng.framework target. I looked around and could find it packaged only as [a sub-framework of a framework that itself depended upon another framework](http://www.kyngchaos.com/software:frameworks). At this point, I gave up, since the changes would be too invasive, no matter what my approach, to be assumed to be acceptable in a blind pull request.
* In addition, building for release would require someone's Developer ID to be signed in a readily distributable way, or require users to change a setting in the Security preference pane to allow apps from all developers.

Compiling with clang did reveal two memory management-related warnings:

* [linker.cpp#246](https://github.com/opensludge/opensludge/blob/e53694f2c0cdb7c82acfcc01b8b724617eddb0b7/source/DevKit/Compiler/linker.cpp#L246): /.../opensludge/source/DevKit/Compiler/linker.cpp:246:2: 'delete' applied to a pointer that was allocated with 'new[]'; did you mean 'delete[]'?
* [splitter.cpp#112](https://github.com/opensludge/opensludge/blob/e53694f2c0cdb7c82acfcc01b8b724617eddb0b7/source/DevKit/Compiler/splitter.cpp#L112): /.../opensludge/source/DevKit/Compiler/splitter.cpp:112:6: 'delete' applied to a pointer that was allocated with 'new[]'; did you mean 'delete[]'?

**Comment by dos1** [comment]
Thanks for looking at it! I'm very interested in OpenSLUDGE's macOS build that includes all the recent fixes, so we can properly release macOS version of https://store.steampowered.com/app/380140/The_Secret_of_Tremendous_Corporation/ :)

Let me find and upload my old patches that port it to SDL2, maybe that'll help a bit :)

**Comment by dos1** [comment]
BTW. Do you have your changes made so far available somewhere in a more structured format? (github fork, patchset, etc. ?)

**Comment by JesperTreetop** [comment]
No, I didn't keep my exact state since I essentially gave up where I stood. I don't have any particular experience in this codebase and don't have the blessing of a maintainer. I thought if it would end up as a relatively clean pull request, I could submit that and it would either be accepted or it wouldn't, but the changes so far seemed too invasive, and everything I know about open source "etiquette" tells me not to do an uproot-the-world pull request without a maintainer telling me what I'm doing is within the realm of the acceptable with regards to the other platforms.

Since it says you're a contributor, I'm open to taking another look at your insistence. I'd wait for the SDL patches you mentioned to land first, if you don't mind.

**Comment by dos1** [comment]
I have found it and just posted #11. Looks like it wasn't a very big effort after all ;D Needs more testing though before it can be merged.

Feel free to post whatever you do as a pull request, even if it's not meant to be merged yet - simply state it clearly in the description :) You can also just push it to your own fork and link it here.

**Comment by JesperTreetop** [comment]
A few weeks ago I made some progress making the Dev Kit build and run and it seems to be working as far as I can see, but I still don't have a good solution for having a libpng framework that will work back to 10.9. Even if I included the current libpng sources as a target and made it build a framework, I don't have access to a machine with 10.9 to test that everything works. The temporary solution I have to see that it's all working is to link a dylib provided by [Homebrew](https://brew.sh/), but they are binary per-OS-version packages, so that won't work for other OS versions or architectures.

I added some new features, including turning on support for the macOS automatic window tabbing if enabled (to open various editing windows as tabs in the project window) and making a floating quick reference panel that shows the HTML documentation of the function under the text insertion point, but none of this is possible to contribute if I don't find a way to ensure that the Dev Kit can run on older (or newer!) versions too.

**Comment by dos1** [comment]
Push it to some branch. I think I have a VM with 10.9 laying somewhere, so I actually might be able to test it.

**Comment by JesperTreetop** [comment]
I managed to build a compatible dylib of libpng from source with a deployment target earlier than 10.9 and have instructions written up. I thought I'd look at the engine now, but there are no instructions whatsoever for non-Linux platforms (including Windows) about how to acquire the dependencies. Among the dependencies is SDL, which means that I'll be waiting until the SDL2 work in #11 has landed. I don't want to do a pull request before I know that I can make both the dev kit and the engine will build cleanly.

**Comment by JesperTreetop** [comment]
I tried today to make some progress to re-do everything based on the recent release and I definitely can't push through trying to get SDL 1.2 to work. There's no pre-bundled framework, and the source has dependencies on X11, which doesn't come with the system. (It's on homebrew, but see a previous comment for why that won't work.) SDL 2 has a pre-bundled framework, so it shouldn't be an obstacle.

I see that GLee was replaced with GLEW. GLEW doesn't seem to work on macOS, but in all cases I can find, everything seems to compile if I just include the system `<OpenGL/GL.h>` header instead (guarded with `#if __APPLE__` of course). This makes sense to me since GLEW is about managing extensions and OpenGL is baked into the system on macOS.

I also noticed that a file was missing preventing the engine from building - see #17. (I never got to the engine before, so this likely isn't new.)

I should note that I am not an expert in C development on macOS, and don't even do it day-to-day. I have spent some time doing Objective-C development, have extremely rudimentary C++ experience (I'm trying not to touch it at all in the code base) and basic C experience. I'm trying to do this mostly because I followed and enjoyed SLUDGE and its games (especially Out of Order) back in the day and I don't want to see OpenSLUDGE fall off the cliff whenever the 32-bit reaper comes around. Someone who knew what they were doing would probably not have such a hard time as I am. Until that someone comes around, my guess is that something being done is probably better than nothing being done, but I could easily be wrong about this.

**Comment by dos1** [comment]
Now that SLUDGE 2.2.2 is released, I'll probably merge the SDL2 port in soon. I've even got some more experience with SDL meanwhile, so making sure everything works fine should be easier now ;)

**Comment by JesperTreetop** [comment]
The next version of macOS, 10.15 Catalina, is due soon and will definitely remove compatibility with 32-bit apps. It doesn't look to me like the SDL2 port is merged yet. I'm ready to take another look when it has been.
